### PR TITLE
Fix chose error materialized view when use count() function

### PR DESCRIPTION
**Pull request by Youngwb**
#427 
1. Fix chose error mv when use count() in query
2. refactor code of  disableAggMV and isSPJQuery for easy understanding

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java [hunk 1]
```diff
@@ -134,9 +146,11 @@ public static boolean isExistMVs(OptExpression root) {
         checkCompensatingPredicates(columnNameToIds.get(tableId), columnIdsInPredicates.get(tableId),
                 candidateIndexIdToMeta);
         // Step3: group by list in query is the subset of group by list in view or view contains no aggregation
-        checkGrouping(columnNameToIds.get(tableId), columnIdsInGrouping.get(tableId), candidateIndexIdToMeta);
+        checkGrouping(columnNameToIds.get(tableId), columnIdsInGrouping.get(tableId), candidateIndexIdToMeta,
+                disableSPJGMVs.get(tableId), isSPJQueries.get(tableId));
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java:150]:
```suggestion
                disableSPJGMVs.get(tableId), isSPJQueries.get(relationId));
```
Can be replace all tableId/scanId/table to relationId?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java [hunk 2]
```diff
@@ -236,11 +250,17 @@ private void collectGroupByAndAggFunctions(OptExpression root) {
         }
 
         Operator operator = root.getOp();
+        if (operator instanceof LogicalOlapScanOperator) {
+            LogicalOlapScanOperator scanOperator = (LogicalOlapScanOperator) operator;
+            traceScanIds.add(factory.getRelationId(scanOperator.getOutputColumns().get(0).getId()));
+        }
+
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java:257]: Why not in collectScanOutputColumns()?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java [hunk 3]
```diff
@@ -64,15 +65,26 @@ public MaterializedViewRule() {
...
     private ColumnRefFactory factory;
-    private boolean disableAggMV = false;
-    private boolean isSPJQuery = false;
+    // record the relation id -> disableSPJGMV flag
+    // this can be set to true when query has count(*) or count(1)
+    private final Map<Integer, Boolean> disableSPJGMVs = Maps.newHashMap();
+    // record the relation id -> isSPJQuery flag
+    private final Map<Integer, Boolean> isSPJQueries = Maps.newHashMap();
+    // record the scan node relation id which has been accessed.
+    private final Set<Integer> traceScanIds = Sets.newHashSet();
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java:74]:
```suggestion
    private final Set<Integer> traceRelationIds = Sets.newHashSet();
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java [hunk 4]
```diff
@@ -269,10 +290,44 @@ private void collectGroupByAndAggFunctions(OptExpression root) {
         }
     }
 
+    // Disable SPJG materialized view for traced scan node which not has aggregation
+    private void disableAggMaterializedView() {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java:294]:
```suggestion
    private void disableSPJGMaterializedView() {
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java (hunk 2 above)
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/mv/MaterializedViewRule.java:257]: collectGroupByAndAggFunctions and  collectScanOutputColumns  both access the entire plan from the bottom up， traceScanIds is used for record relationId below the aggregate node

**Review by Seaven — DISMISSED** [review]

**Review by kangkaisen — DISMISSED** [review]

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**